File: app/src/main/java/com/limelight/grid/AppGridAdapter.java
package com.limelight.grid;

import android.content.Context;
import android.graphics.BitmapFactory;
import android.widget.ImageView;
import android.widget.TextView;

import com.limelight.AppView;
import com.limelight.LimeLog;
import com.limelight.R;
import com.limelight.grid.assets.CachedAppAssetLoader;
import com.limelight.grid.assets.DiskAssetLoader;
import com.limelight.grid.assets.MemoryAssetLoader;
import com.limelight.grid.assets.NetworkAssetLoader;
import com.limelight.nvstream.http.ComputerDetails;
import com.limelight.preferences.PreferenceConfiguration;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

public class AppGridAdapter {
    public interface ChangeListener {
        void onAppsChanged();
    }
    private static final int ART_WIDTH_PX = 300;
    private static final int SMALL_WIDTH_DP = 100;
    private static final int LARGE_WIDTH_DP = 150;

    private final ComputerDetails computer;
    private final String uniqueId;
    private final boolean showHiddenApps;
    private final Context context;

    private CachedAppAssetLoader loader;
    private Set<Integer> hiddenAppIds = new HashSet<>();
    private ArrayList<AppView.AppObject> allApps = new ArrayList<>();
    private final ArrayList<AppView.AppObject> itemList = new ArrayList<>();
    private ChangeListener changeListener;
    private CachedAppAssetLoader.ArtworkLoadListener artworkLoadListener;

    public AppGridAdapter(Context context, PreferenceConfiguration prefs, ComputerDetails computer, String uniqueId, boolean showHiddenApps) {
        this.context = context;
        this.computer = computer;
        this.uniqueId = uniqueId;
        this.showHiddenApps = showHiddenApps;

        updateLayoutWithPreferences(context, prefs);
    }

    public void updateHiddenApps(Set<Integer> newHiddenAppIds, boolean hideImmediately) {
        this.hiddenAppIds.clear();
        this.hiddenAppIds.addAll(newHiddenAppIds);

        if (hideImmediately) {
            // Reconstruct the itemList with the new hidden app set
            itemList.clear();
            for (AppView.AppObject app : allApps) {
                app.isHidden = hiddenAppIds.contains(app.app.getAppId());

                if (!app.isHidden || showHiddenApps) {
                    itemList.add(app);
                }
            }
        }
        else {
            // Just update the isHidden state to show the correct UI indication
            for (AppView.AppObject app : allApps) {
                app.isHidden = hiddenAppIds.contains(app.app.getAppId());
            }
        }

        notifyDataSetChanged();
    }

    public void updateLayoutWithPreferences(Context context, PreferenceConfiguration prefs) {
        int dpi = context.getResources().getDisplayMetrics().densityDpi;
        int dp;

        if (prefs.smallIconMode) {
            dp = SMALL_WIDTH_DP;
        }
        else {
            dp = LARGE_WIDTH_DP;
        }

        double scalingDivisor = ART_WIDTH_PX / (dp * (dpi / 160.0));
        if (scalingDivisor < 1.0) {
            // We don't want to make them bigger before draw-time
            scalingDivisor = 1.0;
        }
        LimeLog.info("Art scaling divisor: " + scalingDivisor);

        if (loader != null) {
            // Cancel operations on the old loader
            cancelQueuedOperations();
        }

        this.loader = new CachedAppAssetLoader(computer, scalingDivisor,
                new NetworkAssetLoader(context, uniqueId),
                new MemoryAssetLoader(),
                new DiskAssetLoader(context),
                BitmapFactory.decodeResource(context.getResources(), R.drawable.no_app_image));
        loader.setArtworkLoadListener(artworkLoadListener);

        notifyDataSetChanged();
    }

    public void cancelQueuedOperations() {
        loader.cancelForegroundLoads();
        loader.cancelBackgroundLoads();
        loader.freeCacheMemory();
    }

    public void setArtworkLoadListener(CachedAppAssetLoader.ArtworkLoadListener listener) {
        artworkLoadListener = listener;
        loader.setArtworkLoadListener(listener);
    }

    public void setChangeListener(ChangeListener listener) {
        changeListener = listener;
    }

    public int getCount() {
        return itemList.size();
    }

    public AppView.AppObject getItem(int position) {
        return itemList.get(position);
    }

    public List<AppView.AppObject> getVisibleApps() {
        return new ArrayList<>(itemList);
    }

    public List<AppView.AppObject> getAllApps() {
        return new ArrayList<>(allApps);
    }

    public void notifyDataSetChanged() {
        if (changeListener != null) {
            changeListener.onAppsChanged();
        }
    }

    private static void sortList(List<AppView.AppObject> list) {
        Collections.sort(list, new Comparator<AppView.AppObject>() {
            @Override
            public int compare(AppView.AppObject lhs, AppView.AppObject rhs) {
                return lhs.app.getAppName().toLowerCase().compareTo(rhs.app.getAppName().toLowerCase());
            }
        });
    }

    public void addApp(AppView.AppObject app) {
        // Update hidden state
        app.isHidden = hiddenAppIds.contains(app.app.getAppId());

        // Always add the app to the all apps list
        allApps.add(app);
        sortList(allApps);

        // Add the app to the adapter data if it's not hidden
        if (showHiddenApps || !app.isHidden) {
            // Queue a request to fetch this bitmap into cache
            loader.queueCacheLoad(app.app);

            // Add the app to our sorted list
            itemList.add(app);
            sortList(itemList);
        }
    }

    public void removeApp(AppView.AppObject app) {
        itemList.remove(app);
        allApps.remove(app);
    }

    public void clear() {
        itemList.clear();
        allApps.clear();
    }

    public void populateArtwork(AppView.AppObject obj, ImageView imageView, TextView textView) {
        loader.populateImageView(obj.app, imageView, textView);
    }
}

File: app/src/main/java/com/limelight/grid/AppGridAdapter.kt
package com.limelight.grid

import android.content.Context
import android.graphics.BitmapFactory
import android.view.View
import android.widget.ImageView
import android.widget.TextView

import com.limelight.AppView
import com.limelight.R
import com.limelight.grid.assets.CachedAppAssetLoader
import com.limelight.grid.assets.DiskAssetLoader
import com.limelight.grid.assets.MemoryAssetLoader
import com.limelight.grid.assets.NetworkAssetLoader
import com.limelight.nvstream.http.ComputerDetails
import com.limelight.preferences.PreferenceConfiguration
import com.limelight.utils.AppIconCache

class AppGridAdapter(
    context: Context,
    prefs: PreferenceConfiguration,
    private val computer: ComputerDetails,
    private val uniqueId: String,
    private val showHiddenApps: Boolean
) : GenericGridAdapter<AppView.AppObject>(context, getLayoutIdForPreferences(prefs)) {

    private var loader: CachedAppAssetLoader? = null
    private var hiddenAppIds: MutableSet<Int> = HashSet()
    private val allApps = ArrayList<AppView.AppObject>()

    init {
        updateLayoutWithPreferences(context, prefs)
    }

    fun updateHiddenApps(newHiddenAppIds: Set<Int>, hideImmediately: Boolean) {
        hiddenAppIds.clear()
        hiddenAppIds.addAll(newHiddenAppIds)

        if (hideImmediately) {
            itemList.clear()
            for (app in allApps) {
                app.isHidden = app.app.appId in hiddenAppIds
                if (!app.isHidden || showHiddenApps) {
                    itemList.add(app)
                }
            }
        } else {
            for (app in allApps) {
                app.isHidden = app.app.appId in hiddenAppIds
            }
        }

        notifyDataSetChanged()
    }

    fun updateLayoutWithPreferences(context: Context, prefs: PreferenceConfiguration) {
        val dpi = context.resources.displayMetrics.densityDpi
        val dp = if (prefs.smallIconMode) SMALL_WIDTH_DP else LARGE_WIDTH_DP

        var scalingDivisor = ART_WIDTH_PX / (dp * (dpi / 160.0))
        if (scalingDivisor < 1.0) {
            scalingDivisor = 1.0
        }

        if (loader != null) {
            cancelQueuedOperations()
        }

        loader = CachedAppAssetLoader(
            context,
            computer,
            scalingDivisor,
            NetworkAssetLoader(context, uniqueId),
            MemoryAssetLoader(),
            DiskAssetLoader(context),
            BitmapFactory.decodeResource(context.resources, R.drawable.no_app_image)
        )

        setLayoutId(getLayoutIdForPreferences(prefs))
    }

    fun cancelQueuedOperations() {
        loader?.cancelForegroundLoads()
        loader?.cancelBackgroundLoads()
        loader?.freeCacheMemory()
    }

    fun getLoader(): CachedAppAssetLoader? = loader

    fun addApp(app: AppView.AppObject) {
        app.isHidden = app.app.appId in hiddenAppIds
        allApps.add(app)

        if (showHiddenApps || !app.isHidden) {
            loader?.queueCacheLoad(app.app)
            itemList.add(app)
        }
    }

    fun removeApp(app: AppView.AppObject) {
        itemList.remove(app)
        allApps.remove(app)
    }

    fun rebuildAppList(newApps: List<AppView.AppObject>) {
        allApps.clear()
        itemList.clear()

        for (app in newApps) {
            app.isHidden = app.app.appId in hiddenAppIds
            allApps.add(app)

            if (showHiddenApps || !app.isHidden) {
                loader?.queueCacheLoad(app.app)
                itemList.add(app)
            }
        }
    }

    override fun clear() {
        super.clear()
        allApps.clear()
    }

    override fun getItemId(i: Int): Long {
        return if (i in 0 until itemList.size) {
            itemList[i].app.appId.toLong()
        } else {
            super.getItemId(i)
        }
    }

    override fun populateView(
        parentView: View,
        imgView: ImageView?,
        spinnerView: View?,
        txtView: TextView?,
        overlayView: ImageView?,
        obj: AppView.AppObject
    ) {
        loader?.populateImageView(obj, imgView!!, txtView, false) { bitmap ->
            AppIconCache.instance.putIcon(computer, obj.app, bitmap)
        }

        if (obj.isRunning) {
            overlayView?.setImageResource(R.drawable.ic_play_cute)
            overlayView?.visibility = View.VISIBLE
        } else {
            overlayView?.visibility = View.GONE
        }

        parentView.alpha = if (obj.isHidden) 0.40f else 1.0f
    }

    companion object {
        private const val ART_WIDTH_PX = 300
        private const val SMALL_WIDTH_DP = 120
        private const val LARGE_WIDTH_DP = 180

        fun getLayoutIdForPreferences(prefs: PreferenceConfiguration): Int {
            return if (prefs.smallIconMode) R.layout.app_grid_item_small else R.layout.app_grid_item
        }

        private fun sortList(list: MutableList<AppView.AppObject>) {
            list.sortWith(Comparator.comparing { it.app.appName.lowercase() })
        }
    }
}
